src/persist-client/src/cache.rs

```diff
@@ -314,19 +367,35 @@ enum StateCacheInit {
 
 impl StateCache {
     /// Returns a new StateCache.
-    pub fn new(metrics: Arc<Metrics>) -> Self {
+    pub fn new(
+        cfg: &PersistConfig,
+        metrics: Arc<Metrics>,
+        pubsub_sender: Arc<dyn PubSubSender>,
+    ) -> Self {
         StateCache {
+            cfg: Arc::new(cfg.clone()),
             metrics,
             states: Default::default(),
+            pubsub_sender,
         }
     }
 
     #[cfg(test)]
     pub(crate) fn new_no_metrics() -> Self {
-        Self::new(Arc::new(Metrics::new(
+        Self::new(
             &PersistConfig::new_for_tests(),
-            &MetricsRegistry::new(),
-        )))
+            Arc::new(Metrics::new(
+                &PersistConfig::new_for_tests(),
+                &MetricsRegistry::new(),
+            )),
+            Arc::new(crate::rpc::NoopPubSubSender),
+        )
+    }
+
+    pub(crate) fn apply_diff(&self, shard_id: &ShardId, diff: VersionedData) {
+        if let Some(state) = self.get_cached(shard_id) {
```

Comment:
in the scramble to get this up I totally forgot this is a mutex call under the hood. yeah lemme poke at this some more

Reply:
okay I came up with an alternative here that's semi-viable, but I think I might leave it for a follow-up PR because it's added complication and I don't think it's necessary while the feature lives behind a flag / is intended for staging. the gist is that we can create a channel/stream of `(ShardId, Weak<dyn DynState>)` from the `StateCache`, which the `subscribe_state_cache_to_pubsub` already has access to, and use it to build a local `HashMap<ShardId, Weak<dyn DynState>>` to avoid any mutex contention.

short-term, as in by tomorrow, I'm leaning towards switching the Mutex to a RwLock and then instrumenting it with some metrics

Reply:
sgtm on all counts

Reply:
okay nvm nerdsnipe very successful, introduced a simpler version than what I mentioned that just fetches a weak ref from the StateCache and holds on to it, so we should be doing only 1 mutex lookup per shard